keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // NodeSelector selects nodes for this node maintenance.
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
+    //
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
+    // DrainPlanEntry podType fields should be in the following order:
+    // nil -> DaemonSet -> Static
+    // DrainPlanEntry priority fields should be in ascending order for each podType.
+    // If the priority and podType are the same, concrete selectors are executed first.
+    //
+    // The following entries are injected into the drainPlan on the NodeMaintenance admission:
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Default"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Default"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "DaemonSet"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "DaemonSet"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Static"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Static"
+    //
+    // Duplicate entries are not allowed.
+    // This field is immutable.
+    DrainPlan []DrainPlanEntry
+
+    // Reason for the maintenance.
+    Reason string
+}
+
+const (
+// Default selects all pods except DaemonSet and Static pods.	
+Default PodType = "Default"
+// DaemonSet selects DaemonSet pods.
+DaemonSet PodType = "DaemonSet"
+// Static selects static pods.
+Static PodType = "Static"
+)
+
+type DrainPlanEntry struct {
+    // PodSelector selects pods according to their labels.
+    // This can help to select which pods of the same priority should be evacuated first.
+    // +optional
+    PodSelector *metav1.LabelSelector
```

Comment:
Should it also be possible to select by namespace?

Reply:
IMO: not for alpha!

When you maintain a node, you expect to take all the Pods offline or put them at risk. Let's ship something useful and then see if we want to iterate.

Reply:
Begin this a global CRD, the lack of a namespace selector seems a relevant GAP for adoption

Reply:
What is the intended use case and benefit for the namespace? 

The main differentiator should be the pod priority. I have added the label selector to mainly differentiate between different components running for example the ` system-cluster-critical` or the `system-node-critical` priority. For example, if you are running some special etcd/apiserver configuration, you could control how the instances downscale during an upgrade. So, basically to have an additional ordering to just a priority. But in most cases the priority should be enough, and downscaling by namespaces could slow down the process considerably.

Reply:
To me it seems unrealistic that ordering for Pod drain will be configured by setting Pod priorities on all Pods in a cluster, especially if not a single team has control over an entire cluster and can set Pod priorities on all relevant Pods.

We had this case here in Cluster API where folks where using Portworx which uses an operator to manage Pods: https://github.com/kubernetes-sigs/cluster-api/issues/11024 (not a DaemonSet)

I assume that if there is a use-case for a PodSelector the same probably also applies for a namespace selector (I think outside of very limited use cases it's not safe to assume Pod labels are unique across namespaces)